cmd/oras/manifest/push.go

```diff
@@ -0,0 +1,153 @@
+limitations under the License.
+*/
+
+package manifest
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras/cmd/oras/internal/display"
+	"oras.land/oras/cmd/oras/internal/option"
+	"oras.land/oras/internal/file"
+)
+
+type pushOptions struct {
+	option.Common
+	option.Descriptor
+	option.Pretty
+	option.Remote
+
+	targetRef string
+	fileRef   string
+	mediaType string
+}
+
+func pushCmd() *cobra.Command {
+	var opts pushOptions
+	cmd := &cobra.Command{
+		Use:   "push [flags] name[:tag|@digest] file",
+		Short: "[Preview] Push a manifest to remote registry",
+		Long: `[Preview] Push a manifest to remote registry
+
+** This command is in preview and under development. **
+
+Example - Push a manifest to repository 'locahost:5000/hello' and tag with 'latest':
+  oras manifest push localhost:5000/hello:latest manifest.json
+
+Example - Push a manifest with content read from stdin:
+  oras manifest push localhost:5000/hello:latest -
+
+Example - Push a manifest and output its descriptor:
+  oras manifest push --descriptor localhost:5000/hello:latest manifest.json
+
+Example - Push a manifest to repository 'locahost:5000/hello' and output the prettified descriptor:
+  oras manifest push --descriptor --pretty localhost:5000/hello manifest.json
+
+Example - Push a manifest with specified media type to repository 'locahost:5000/hello' and tag with 'latest':
+  oras manifest push --media-type application/vnd.cncf.oras.artifact.manifest.v1+json localhost:5000/hello:latest oras_manifest.json
+`,
+		Args: cobra.ExactArgs(2),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			if opts.fileRef == "-" && opts.PasswordFromStdin {
+				return errors.New("`-` read file from input and `--password-stdin` read password from input cannot be both used")
+			}
+			return opts.ReadPassword()
+		},
+		RunE: func(_ *cobra.Command, args []string) error {
+			opts.targetRef = args[0]
+			opts.fileRef = args[1]
+			return pushManifest(opts)
+		},
+	}
+
+	option.ApplyFlags(&opts, cmd.Flags())
+	cmd.Flags().StringVarP(&opts.mediaType, "media-type", "", "", "media type of manifest")
+	return cmd
+}
+
+func pushManifest(opts pushOptions) error {
+	ctx, _ := opts.SetLoggerLevel()
+	repo, err := opts.NewRepository(opts.targetRef, opts.Common)
+	if err != nil {
+		return err
+	}
+	manifests := repo.Manifests()
+
+	// prepare manifest content
+	contentBytes, err := file.PrepareManifestContent(opts.fileRef)
+	if err != nil {
+		return err
+	}
+
+	// get manifest media type
+	mediaType := opts.mediaType
+	if opts.mediaType == "" {
+		mediaType, err = file.ParseMediaType(contentBytes)
+		if err != nil {
+			return err
+		}
+	}
+
+	// prepare manifest descriptor
+	desc := content.NewDescriptorFromBytes(mediaType, contentBytes)
+
+	var ref string
+	if tag := repo.Reference.Reference; tag != "" {
+		ref = tag
+	} else {
+		ref = desc.Digest.String()
+	}
```

Comment:
nit: code block can be simplified.
```suggestion
	ref := repo.Reference.Reference
	if ref == "" {
		ref = desc.Digest.String()
	}
```